### DISCUSSION TITLE
Error Assigning Elasticity Tensor with AD Material

### DISCUSSION THREAD

**Discussion by m1russell** [idaholab/moose#27463] [Q&A Modules: Solid mechanics]
I have ran into the following error while attempting to convert a material class for thermally varying orthotopic elasticity to be compatible with AD using the "Supporting both AD and non-AD variables through templating" guide.   The error occurs at the  line where I assign my rank four orthroptic tesnor to the elasticity tensor:

`  _elasticity_tensor[_qp] = _ortho_Cijkl;/`

The error message is as follows:
`/projects/moose/framework/build/header_symlinks/RankFourTensor.h: In instantiation of 'RankFourTensorTempl< <template-parameter-1-1> >::RankFourTensorTempl(const RankFourTensorTempl<T2>&) [with T2 = MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true>; T = double]':
/projects/phoenix/src/materials/ComputeVariableOrthotropicElasticityTensor.C:132:29:   required from 'void ComputeVariableOrthotropicElasticityTensorTempl<is_ad>::computeQpElasticityTensor() [with bool is_ad = true]'
/projects/phoenix/src/materials/ComputeVariableOrthotropicElasticityTensor.C:139:16:   required from here
/projects/moose/framework/build/header_symlinks/RankFourTensor.h:544:26: error: cannot convert 'const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true>' to 'double' in assignment
  544 |     _vals[i] = copy._vals[i];
      |                ~~~~~~~~~~^
      |                          |
      |                          const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true>
`

I attempted to use Generic type variable declaration but perhaps I got something wrong? 

Below are my computeQpElasticityTensor function and .H file:

```
template <bool is_ad>
void
ComputeVariableOrthotropicElasticityTensorTempl<is_ad>::computeQpElasticityTensor()
{
  
    _ortho_stiffness_components.resize(12);

    GenericReal<is_ad> nu21 = (_nu12[_qp] * _E22[_qp]) / _E11[_qp];
    GenericReal<is_ad> nu31 = (_nu13[_qp] * _E33[_qp]) / _E11[_qp] ;
    GenericReal<is_ad> nu32 = (_nu23[_qp] * _E33[_qp]) / _E22[_qp] ;

    _ortho_stiffness_components[0]  = _E11[_qp] ;
    _ortho_stiffness_components[1]  = _E22[_qp] ;
    _ortho_stiffness_components[2]  = _E33[_qp] ;
    _ortho_stiffness_components[3]  = _G12[_qp] ;
    _ortho_stiffness_components[4]  = _G23[_qp] ;
    _ortho_stiffness_components[5]  = _G13[_qp] ;
    _ortho_stiffness_components[6]  = nu21;
    _ortho_stiffness_components[7]  = nu31; 
    _ortho_stiffness_components[8]  = nu32;
    _ortho_stiffness_components[9]  = _nu12[_qp] ;
    _ortho_stiffness_components[10]  = _nu13[_qp] ; 
    _ortho_stiffness_components[11]  = _nu23[_qp] ;


    //input	Ea, Eb, Ec, Gab, Gbc, Gca, nuba, nuca, nucb, nuab, nuac, nubc
    _ortho_Cijkl.fillFromInputVector(_ortho_stiffness_components,  GenericRankFourTensor<is_ad> :: orthotropic);

    if(_read_prop_user_object)
    {
      _ortho_Cijkl.rotate(_rotation_matrix_from_file_old[_qp]);//use _old to allow use of initQpStatefulProperties to only read once
    }
    else
    {
      _ortho_Cijkl.rotate(_rotation_matrix);//_rotation_matrix intialized a identity, can probably avoid rotation if not defined as a oarameter
    }
    // Assign elasticity tensor at a given quad point
    _elasticity_tensor[_qp] = _ortho_Cijkl;//used in material/ComputeLinearElasticStress.C

    

}

template class ComputeVariableOrthotropicElasticityTensorTempl<false>;
template class ComputeVariableOrthotropicElasticityTensorTempl<true>;

```

`
```
protected:
  virtual void initialSetup() override;
  virtual void initQpStatefulProperties() override;
  virtual void computeQpElasticityTensor() override;

  
  /// vector to store the stiffness values for populating matrix
  std::vector<GenericReal<is_ad>> _ortho_stiffness_components ; 
  /// matrix to store the stiffness values
  GenericRankFourTensor<is_ad>  _ortho_Cijkl ;

  /// Material defining the Modulli
  const GenericMaterialProperty<Real, is_ad>& _E11;
  const GenericMaterialProperty<Real, is_ad>& _E22;
  const GenericMaterialProperty<Real, is_ad>& _E33;
  const GenericMaterialProperty<Real, is_ad>& _G12;
  const GenericMaterialProperty<Real, is_ad>& _G13;
  const GenericMaterialProperty<Real, is_ad>& _G23;
  const GenericMaterialProperty<Real, is_ad>& _nu12;
  const GenericMaterialProperty<Real, is_ad>& _nu13;
  const GenericMaterialProperty<Real, is_ad>& _nu23;
  
  
  /// number of variables the moduli depend on
  const unsigned int _num_args;

  ///Element property read user object used to read in Euler angles
  const PropertyReadFile * const _read_prop_user_object;

  /// rotation Matrix Read in from file
  GenericMaterialProperty<RankTwoTensor, is_ad> & _rotation_matrix_from_file;
  const GenericMaterialProperty<RankTwoTensor, is_ad> & _rotation_matrix_from_file_old;
```

};

typedef ComputeVariableOrthotropicElasticityTensorTempl<false> ComputeVariableOrthotropicElasticityTensor;
typedef ComputeVariableOrthotropicElasticityTensorTempl<true> ADComputeVariableOrthotropicElasticityTensor;
`

**Comment by GiudGiud** [comment]
Hello

The compiler is preventing you from losing the AD information by converting to 'double' from an 'ADReal'
```
cannot convert 'const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true>' to 'double'
```

if you really mean to do that (it can be fair sometimes) then use `MetaPhysicL::raw_value(` on the number to turn it from AD to double

Guillaume
